### DISCUSSION TITLE
Easy development of SPIs

### DISCUSSION THREAD

**Discussion by taypo** [keycloak/keycloak#20864] [Archive]
I need to develop an authorization SPI extension. Is it possible to run Keycloak with my SPI inside the IDE, and not include keycloak source code in my repository. I just want to add it as a maven dependency and run/debug. I've read previous discussions that this is on the roadmap last year. Couldn't find if its done or not.

Thank you

**Comment by bernhof** [comment]
If you run Keycloak with your custom SPI in a container, you can set the following environment variables to allow remote JVM debugging:

```
DEBUG=true
DEBUG_PORT=*:8787
```

Remember to expose port 8787 on the container when running it.

This should allow you to set breakpoints and debug your code as usual.

**Comment by martinleim** [comment]
Hi @taypo , you might want to have a look at https://github.com/dasniko/testcontainers-keycloak, which allows testing custom Keycloak SPIs in Docker containers. It also supports remote debugging, like it has suggested in the previous post.
